tools/bridge/end2end-tests/tests/test_transfer.py

```diff
@@ -0,0 +1,427 @@
+import os
+import subprocess
+import time
+import warnings
+from subprocess import Popen
+
+import pytest
+import requests
+from deploy_tools import deploy_compiled_contract
+from deploy_tools.deploy import wait_for_successful_transaction_receipt
+from web3 import HTTPProvider, Web3
+from web3.contract import Contract
+from web3.middleware import construct_sign_and_send_raw_middleware
+
+PARITY_DEV_ACCOUNT = "0x00a329c0648769A73afAc7F9381E08FB43dBEA72"
+PARITY_DEV_KEY = "0x4D5DB4107D237DF6A3D58EE5F70AE63D73D7658D4026F2EEFD2F204C81682CB7"
+
+HOME_RPC_URL = "http://localhost:8545"
+FOREIGN_RPC_URL = "http://localhost:8645"
+
+START_WEI_PER_ACCOUNT = 1 * 10 ** 18
+BRIDGE_FUNDS = 1_000_000 * 10 ** 18
+
+
+class TimeoutException(Exception):
+    pass
+
+
+class Timer:
+    def __init__(self, timeout):
+        self.start_time = None
+        self.timeout = timeout
+
+    def start(self):
+        self.start_time = time.time()
+
+    def is_timed_out(self):
+        if self.start_time is None:
+            raise ValueError("Timer is not started yet")
+        return self.time_passed > self.timeout
+
+    @property
+    def time_passed(self):
+        return time.time() - self.start_time
+
+
+def assert_within_timeout(check_function, timeout, poll_period=0.5):
+    """
+    Runs a check_function with assertions and give it some time to pass
+    :param check_function: The function which will be periodically called. It should contain some assertions
+    :param timeout: After this timeout non passing assertion will be raised
+    :param poll_period: poll interval to check the check_function
+    """
+    timer = Timer(timeout)
+    timer.start()
+
+    while True:
+        try:
+            check_function()
+        except AssertionError as e:
+            if not timer.is_timed_out():
+                time.sleep(poll_period)
```

Comment:
Nitpick:
```suggestion
                time.sleep(max(poll_period, timer.time_left))
```

where `time_left` is a to-be-added property. Otherwise we might not time out even if we should.

Reply:
done